utils/proc/process.go

```
@@ -0,0 +1,208 @@
+}
+
+func (p *ps) PPid() int {
+	ppid, _ := p.imp.Ppid()
+	return int(ppid)
+}
+
+func (p *ps) Executable() string {
+	ex, _ := p.imp.Exe()
+	return ex
+}
+
+func (p *ps) Terminate(ctx context.Context) error {
+	err := ConvertProcessError(p.imp.TerminateWithContext(ctx))
+	if commonerrors.Any(err, commonerrors.ErrNotFound) {
+		err = nil
+	}
+	return err
+}
+
+func (p *ps) KillWithChildren(ctx context.Context) error {
+	return killProcessAndChildren(ctx, p.imp)
+}
+
+func killProcessAndChildren(ctx context.Context, p *process.Process) (err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	if p == nil {
+		return
+	}
+	defer func() {
+		_ = p.Kill()
+	}()
+	err = ConvertProcessError(p.TerminateWithContext(ctx))
+	if err != nil {
+		if commonerrors.Any(err, commonerrors.ErrNotFound) {
+			err = nil
+		}
+		return
+	}
+	err = ConvertProcessError(killGroup(ctx, p.Pid))
+	if err == nil {
+		if commonerrors.Any(err, commonerrors.ErrNotFound) {
+			err = nil
+		}
+		return
+	}
+	err = killChildren(ctx, p)
+	if err != nil {
+		if commonerrors.Any(err, commonerrors.ErrNotFound) {
+			err = nil
+		}
+		return
+	}
+	err = ConvertProcessError(p.KillWithContext(ctx))
+	if commonerrors.Any(err, commonerrors.ErrNotFound) {
+		err = nil
+	}
+	return
+}
+
+func killChildren(ctx context.Context, p *process.Process) (err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	if p == nil {
+		return
+	}
+	children, childErr := p.ChildrenWithContext(ctx)
+	if childErr != nil {
+		childErr = ConvertProcessError(childErr)
+		if commonerrors.Any(childErr, commonerrors.ErrTimeout, commonerrors.ErrCancelled) {
+			err = childErr
+		}
+		return
+	}
+	for i := range children {
+		subErr := killProcessAndChildren(ctx, children[i])
+		if subErr != nil {
+			err = subErr
+		}
+	}
+	if err != nil {
+		_ = killGroup(ctx, p.Pid) // Radical approach of killing the whole process group
+	}
+	return err
```

do you still want to return the error from before the `killGroup`? This will mean the function errors even if `killGroup` succeeds

---

fair point if no error is returned I will set err to nil